// JQDeferred/test/deferred.js
var Deferred = require( "../lib/jquery-deferred" ),
	jQuery = require( "../lib/jquery" );

jQuery.each( [ "", " - new operator" ], function( _, withNew ) {

	function createDeferred( fn ) {
		return withNew ? new Deferred( fn ) : Deferred( fn );
	}

	exports[ "Deferred" + withNew ] = (function( ______ ) {

		______.expect( 22 );

		createDeferred().resolve().then( function() {
			______.ok( true , "Success on resolve" );
			______.ok( this.isResolved(), "Deferred is resolved" );
			______.strictEqual( this.state(), "resolved", "Deferred is resolved (state)" );
		}, function() {
			______.ok( false , "Error on resolve" );
		}).always( function() {
			______.ok( true , "Always callback on resolve" );
		});

		createDeferred().reject().then( function() {
			______.ok( false , "Success on reject" );
		}, function() {
			______.ok( true , "Error on reject" );
			______.ok( this.isRejected(), "Deferred is rejected" );
			______.strictEqual( this.state(), "rejected", "Deferred is rejected (state)" );
		}).always( function() {
			______.ok( true , "Always callback on reject" );
		});

		createDeferred( function( defer ) {
			______.ok( this === defer , "Defer passed as this & first argument" );
			this.resolve( "done" );
		}).then( function( value ) {
			______.strictEqual( value , "done" , "Passed function executed" );
		});

		jQuery.each( "resolve reject".split( " " ), function( _, change ) {
			createDeferred( function( defer ) {
				______.strictEqual( defer.state(), "pending", "pending after creation" );
				var checked = 0;
				defer.progress(function( value ) {
					______.strictEqual( value, checked, "Progress: right value (" + value + ") received" );
				});
				for( checked = 0; checked < 3 ; checked++ ) {
					defer.notify( checked );
				}
				______.strictEqual( defer.state(), "pending", "pending after notification" );
				defer[ change ]();
				______.notStrictEqual( defer.state(), "pending", "not pending after " + change );
				defer.notify();
			});
		});

		______.done();
	});
} );

exports[ "Deferred - chainability" ] = (function( ______ ) {

	var methods = "resolve reject notify resolveWith rejectWith notifyWith done fail progress then always".split( " " ),
		defer = Deferred();

	______.expect( methods.length );

	jQuery.each( methods, function( _, method ) {
		var object = { m: defer[ method ] };
		______.strictEqual( object.m(), object, method + " is chainable" );
	});

	______.done();
});

exports[ "Deferred.pipe - filtering (done)" ] = (function( ______ ) {

	______.expect(4);

	var defer = Deferred(),
		piped = defer.pipe(function( a, b ) {
			return a * b;
		}),
		value1,
		value2,
		value3;

	piped.done(function( result ) {
		value3 = result;
	});

	defer.done(function( a, b ) {
		value1 = a;
		value2 = b;
	});

	defer.resolve( 2, 3 );

	______.strictEqual( value1, 2, "first resolve value ok" );
	______.strictEqual( value2, 3, "second resolve value ok" );
	______.strictEqual( value3, 6, "result of filter ok" );

	Deferred().reject().pipe(function() {
		______.ok( false, "pipe should not be called on reject" );
	});

	Deferred().resolve().pipe( jQuery.noop ).done(function( value ) {
		______.strictEqual( value, undefined, "pipe done callback can return undefined/null" );
	});

	______.done();
});

exports[ "Deferred.pipe - filtering (fail)" ] = (function( ______ ) {

	______.expect(4);

	var defer = Deferred(),
		piped = defer.pipe( null, function( a, b ) {
			return a * b;
		} ),
		value1,
		value2,
		value3;

	piped.fail(function( result ) {
		value3 = result;
	});

	defer.fail(function( a, b ) {
		value1 = a;
		value2 = b;
	});

	defer.reject( 2, 3 );

	______.strictEqual( value1, 2, "first reject value ok" );
	______.strictEqual( value2, 3, "second reject value ok" );
	______.strictEqual( value3, 6, "result of filter ok" );

	Deferred().resolve().pipe( null, function() {
		______.ok( false, "pipe should not be called on resolve" );
	} );

	Deferred().reject().pipe( null, jQuery.noop ).fail(function( value ) {
		______.strictEqual( value, undefined, "pipe fail callback can return undefined/null" );
	});

	______.done();
});

exports[ "Deferred.pipe - filtering (progress)" ] = (function( ______ ) {

	______.expect(3);

	var defer = Deferred(),
		piped = defer.pipe( null, null, function( a, b ) {
			return a * b;
		} ),
		value1,
		value2,
		value3;

	piped.progress(function( result ) {
		value3 = result;
	});

	defer.progress(function( a, b ) {
		value1 = a;
		value2 = b;
	});

	defer.notify( 2, 3 );

	______.strictEqual( value1, 2, "first progress value ok" );
	______.strictEqual( value2, 3, "second progress value ok" );
	______.strictEqual( value3, 6, "result of filter ok" );

	______.done();
});

exports[ "Deferred.pipe - deferred (done)" ] = (function( ______ ) {

	______.expect(3);

	var defer = Deferred(),
		piped = defer.pipe(function( a, b ) {
			return Deferred(function( defer ) {
				defer.reject( a * b );
			});
		}),
		value1,
		value2,
		value3;

	piped.fail(function( result ) {
		value3 = result;
	});

	defer.done(function( a, b ) {
		value1 = a;
		value2 = b;
	});

	defer.resolve( 2, 3 );

	______.strictEqual( value1, 2, "first resolve value ok" );
	______.strictEqual( value2, 3, "second resolve value ok" );
	______.strictEqual( value3, 6, "result of filter ok" );

	______.done();
});

exports[ "Deferred.pipe - deferred (fail)" ] = (function( ______ ) {

	______.expect(3);

	var defer = Deferred(),
		piped = defer.pipe( null, function( a, b ) {
			return Deferred(function( defer ) {
				defer.resolve( a * b );
			});
		} ),
		value1,
		value2,
		value3;

	piped.done(function( result ) {
		value3 = result;
	});

	defer.fail(function( a, b ) {
		value1 = a;
		value2 = b;
	});

	defer.reject( 2, 3 );

	______.strictEqual( value1, 2, "first reject value ok" );
	______.strictEqual( value2, 3, "second reject value ok" );
	______.strictEqual( value3, 6, "result of filter ok" );

	______.done();
});

exports[ "Deferred.pipe - deferred (progress)" ] = (function( ______ ) {

	______.expect(3);

	var defer = Deferred(),
		piped = defer.pipe( null, null, function( a, b ) {
			return Deferred(function( defer ) {
				defer.resolve( a * b );
			});
		} ),
		value1,
		value2,
		value3;

	piped.done(function( result ) {
		value3 = result;
	});

	defer.progress(function( a, b ) {
		value1 = a;
		value2 = b;
	});

	defer.notify( 2, 3 );

	______.strictEqual( value1, 2, "first progress value ok" );
	______.strictEqual( value2, 3, "second progress value ok" );
	______.strictEqual( value3, 6, "result of filter ok" );

	______.done();
});

exports[ "Deferred.pipe - context" ] = (function( ______ ) {

	______.expect(4);

	var context = {};

	Deferred().resolveWith( context, [ 2 ] ).pipe(function( value ) {
		return value * 3;
	}).done(function( value ) {
		______.strictEqual( this, context, "custom context correctly propagated" );
		______.strictEqual( value, 6, "proper value received" );
	});

	var defer = Deferred(),
		piped = defer.pipe(function( value ) {
			return value * 3;
		});

	defer.resolve( 2 );

	piped.done(function( value ) {
		______.strictEqual( this.promise(), piped, "default context gets updated to latest defer in the chain" );
		______.strictEqual( value, 6, "proper value received" );
	});

	______.done();
});

exports[ "Deferred.when" ] = (function( ______ ) {

	______.expect( 23 );

	// Some other objects
	jQuery.each( {

		"an empty string": "",
		"a non-empty string": "some string",
		"zero": 0,
		"a number other than zero": 1,
		"true": true,
		"false": false,
		"null": null,
		"undefined": undefined,
		"a plain object": {}

	} , function( message , value ) {

		______.ok( "function" === jQuery.type( Deferred.when( value ).done(function( resolveValue ) {
			______.strictEqual( resolveValue , value , "Test the promise was resolved with " + message );
		}).promise ) , "Test " + message + " triggers the creation of a new Promise" );

	} );

	______.ok( "function" === jQuery.type( Deferred.when().done(function( resolveValue ) {
		______.strictEqual( resolveValue , undefined , "Test the promise was resolved with no parameter" );
	}).promise ) , "Test calling when with no parameter triggers the creation of a new Promise" );

	var cache, i;

	for( i = 1 ; i < 4 ; i++ ) {
		Deferred.when( cache || Deferred( function() {
			this.resolve( i );
		}) ).done(function( value ) {
			______.strictEqual( value , 1 , "Function executed" + ( i > 1 ? " only once" : "" ) );
			cache = value;
		});
	}

	______.done();
});

exports[ "Deferred.when - joined" ] = (function( ______ ) {

	______.expect(53);

	var deferreds = {
			value: 1,
			success: Deferred().resolve( 1 ),
			error: Deferred().reject( 0 ),
			futureSuccess: Deferred().notify( true ),
			futureError: Deferred().notify( true ),
			notify: Deferred().notify( true )
		},
		willSucceed = {
			value: true,
			success: true,
			futureSuccess: true
		},
		willError = {
			error: true,
			futureError: true
		},
		willNotify = {
			futureSuccess: true,
			futureError: true,
			notify: true
		};

	jQuery.each( deferreds, function( id1, defer1 ) {
		jQuery.each( deferreds, function( id2, defer2 ) {
			var shouldResolve = willSucceed[ id1 ] && willSucceed[ id2 ],
				shouldError = willError[ id1 ] || willError[ id2 ],
				shouldNotify = willNotify[ id1 ] || willNotify[ id2 ],
				expected = shouldResolve ? [ 1, 1 ] : [ 0, undefined ],
			    expectedNotify = shouldNotify && [ willNotify[ id1 ], willNotify[ id2 ] ],
			    code = id1 + "/" + id2;

			var promise = Deferred.when( defer1, defer2 ).done(function( a, b ) {
				if ( shouldResolve ) {
					______.deepEqual( [ a, b ], expected, code + " => resolve" );
				} else {
					______.ok( false ,  code + " => resolve" );
				}
			}).fail(function( a, b ) {
				if ( shouldError ) {
					______.deepEqual( [ a, b ], expected, code + " => reject" );
				} else {
					______.ok( false ,  code + " => reject" );
				}
			}).progress(function progress( a, b ) {
				______.deepEqual( [ a, b ], expectedNotify, code + " => progress" );
			});
		} );
	} );
	deferreds.futureSuccess.resolve( 1 );
	deferreds.futureError.reject( 0 );

	______.done();
});
